rust/utils/run-locally/src/cosmos/source.rs

```diff
@@ -0,0 +1,153 @@
+use std::{collections::BTreeMap, fs, path::PathBuf};
+
+use tempfile::tempdir;
+
+use crate::{
+    cosmos::{
+        make_target,
+        utils::{download, untar},
+    },
+    logging::log,
+    utils::concat_path,
+};
+
+use super::{CW_HYPERLANE_GIT, CW_HYPERLANE_VERSION, OSMOSIS_CLI_GIT, OSMOSIS_CLI_VERSION};
+
+pub enum CodeSource {
+    Local { path: String },
+    Remote { url: String, version: String },
+}
+
+impl Default for CodeSource {
+    fn default() -> Self {
+        Self::remote(CW_HYPERLANE_GIT, CW_HYPERLANE_VERSION)
+    }
+}
+
+impl CodeSource {
+    pub fn local(path: &str) -> Self {
+        Self::Local {
+            path: path.to_string(),
+        }
+    }
+
+    pub fn remote(url: &str, version: &str) -> Self {
+        Self::Remote {
+            url: url.to_string(),
+            version: version.to_string(),
+        }
+    }
+}
+
+impl CodeSource {
+    fn install_local(src: String) -> BTreeMap<String, PathBuf> {
+        // make contract_name => path map
```

Comment:
this comment is present repeat three times across the project, and the code it describes looks identical each time. Let's try reusing this function everywhere instead of duplicating